src/db/pool.cr

```diff
@@ -34,24 +38,79 @@ module DB
     @availability_channel : Channel(Nil)
     # signal how many existing connections are waited for
     @waiting_resource : Int32
+    # communicate when the pool is closed, to kill the Connection Reaper
+    @close_channel : Channel(Nil)
     # global pool mutex
     @mutex : Mutex
 
-    def initialize(@initial_pool_size = 1, @max_pool_size = 0, @max_idle_pool_size = 1, @checkout_timeout = 5.0,
-                   @retry_attempts = 1, @retry_delay = 0.2, &@factory : -> T)
+    def initialize(
+      @initial_pool_size = 1,
+      @max_pool_size = 0,
+      @max_idle_pool_size = 1,
+      @checkout_timeout = 5.0,
+      @retry_attempts = 1,
+      @retry_delay = 0.2,
+      @reaping_delay = 60.0,
+      @reaping_frequency = 60.0,
+      &@factory : -> T
+    )
       @availability_channel = Channel(Nil).new
+      @close_channel = Channel(Nil).new
       @waiting_resource = 0
       @inflight = 0
       @mutex = Mutex.new
 
       @initial_pool_size.times { build_resource }
     end
 
+    def reap_connection!(connection : T, delay : Float64)
+      sleep delay
+
+      sync do
+        return unless @idle.includes? connection
+
+        delete connection
+      end
+
+      connection.check
+
+      sync do
+        @total << connection
+        @idle << connection
+      end
+    rescue exc : PoolResourceLost(T)
+      connection.close
+      sync { add_resource! if can_increase_pool? }
```

Comment:
Does this rescue block need to span the entire method? I suppose it could only cover the `connection.check` call, right?

Reply:
The connection needs to be added back to @total and @idle on success, so I can't quite see an alternative way to write the rescue to only span the `connection.check` call -- but definitely open to ideas as the only area where PoolResourceLost would be raised is in the connection.close call.

Reply:
@straight-shoota LMK if you have a suggestion for this one. 👍